lib/jsdom/living/events/MouseEvent-impl.js

```diff
@@ -16,31 +16,43 @@ class MouseEventImpl extends UIEventImpl {
   }
   get pageX() {
     if (this._dispatchFlag) {
-      return 0;
+      return this._initialContainingBlockRelativeX;
     }
     const offset = wrapperForImpl(this.view)?.scrollX || 0;
     return offset + this.clientX;
   }
   get pageY() {
     if (this._dispatchFlag) {
-      return 0;
+      return this._initialContainingBlockRelativeY;
     }
     const offset = wrapperForImpl(this.view)?.scrollY || 0;
     return offset + this.clientY;
   }
   get offsetX() {
     if (this._dispatchFlag) {
-      return 0;
+      return this._paddingEdgeRelativeX;
     }
     return this.pageX;
   }
   get offsetY() {
     if (this._dispatchFlag) {
-      return 0;
+      return this._paddingEdgeRelativeY;
     }
     return this.pageY;
   }
 
+  constructor(globalObject, args, privateData) {
+    super(globalObject, args, privateData);
+
+    // We cache this during init so that we have a stable value during dispatch
+    const { scrollX = 0, scrollY = 0 } = wrapperForImpl(this.view) ?? {};
+    this._initialContainingBlockRelativeX = this.clientX + scrollX;
+    this._initialContainingBlockRelativeY = this.clientY + scrollY;
+    // TODO: support non-zero values somehow if/when jsdom adds layout support
+    this._paddingEdgeRelativeX = 0;
+    this._paddingEdgeRelativeY = 0;
```

Comment:
If the aim is to mimic Chrome's implementation details, then maybe? But I'm not sure how how that changes the end result, and IMO it would be less clear at a glance that the code is following the spec. There are some other things happening in the Chrome implementation which I'm not sure we want to follow, e.g. lazily resolving coordinates when first access, which is [not without problems](https://github.com/w3c/csswg-drafts/issues/1070#issuecomment-715590900).

It seems that "having a target" is Chrome's proxy for "is the dispatch flag set?", so it appears the only time the `pageX/Y -> offsetX/Y` defaults get used is outside of dispatch, but the implementation appears inconsistent with the spec and others (see my other comment).

Considering all that, `0` feels like a reasonable initial value for `_offsetX/Y` IMO. If/when jsdom adds layout support, we could then of course implement similar logic to Chrome/others.

Reply:
> Considering all that, `0` feels like a reasonable initial value for `offsetX/Y` IMO.

Seems we both have different criteria about what "reasonable" means, so let me explain how I'm thinking about this 🙂

The `0` value maybe satisfies the spec, but in my view that's not enough. The JSDOM's `MouseEvent` implementation must also serve the practical purposes for which people are using it. And one of the main practical purposes of JSDOM is unit testing of web code, like React components. The reason why I came here to discuss `MouseEvent` is that after upgrading JSDOM to the latest version, one of our unit tests started failing. The test uses the [`react-colorful`](https://omgovich.github.io/react-colorful/) library for implementing color sliders like this:

<img width="221" alt="Screenshot 2023-03-06 at 8 18 10" src="https://user-images.githubusercontent.com/664258/223045091-9bed4c99-ba20-46a1-95a9-b29ec999b127.png">

And the test creates and dispatches `mousedown` and `mousemove` events to simulate moving the slider. On `mousemove`, the tested component calculates relative position in the `getRelativePosition` [method](https://github.com/omgovich/react-colorful/blob/master/src/components/common/Interactive.tsx#L39-L40), reading the `.pageX` and `.pageY` fields. It would be probably better off reading `.offsetX` and `.offsetY`, but anyway. Because JSDOM doesn't support layout, and `getBoundingClientRect` always returns zero height and width, the test is very fragile, but it can at least detect that the slider has moved. If `.pageX` starts always returning `0`, the move will not be detected: the mouse position is always `(0, 0)`!

That's why I ask that `.pageX` and `.offsetX` don't completely ignore the user-specified `.clientX`, but that they should be derived from it. One day, when JSDOM implements even the most minimal layout support, like `getBoundingClientRect` looking at the `width` and `height` CSS styles, the test will become much closer to the browser environment.

Reply:
This is a really illustrative example, thank you!

I think we we were already basically in agreement about `pageX/Y` (thus this PR 😉), so it seems like the problem you describe would mostly be covered by this fix, no? That said, you do make a compelling case for `offsetX/Y` simply mirroring `pageX/Y` in the dispatch case too, in the absence of layout support.

I had been viewing this behavior more from the use case of `click` events on things like image buttons, rather than for scenarios like yours (e.g. detecting mouse movement over time). My concern was primarily around `offsetX/Y` returning surprising values during the former, but upon further reflection that is unlikely to be an issue unless `clientX/Y` or `scrollX/Y` are also explicitly set; typically they wouldn't be for these use cases, so the values will continue to be zero.

Although neither approach fully follows the spec, yours does get us marginally closer and helps solve the problems you've outlined, so I'm on board, sorry it took me a bit to get there 😆